R/data-converters.R

```diff
@@ -0,0 +1,308 @@
+    counts[, reference_date := get(reference_date)]
+  }
+  counts <- coerce_dt(counts, dates = TRUE, copy = FALSE)
+
+  data.table::setkeyv(counts, c(by, "reference_date", "report_date"))
+  counts <- counts[,
+    .(new_confirm = .N), keyby = c("reference_date", "report_date", by)
+  ]
+
+  obs_delay <- max(counts$report_date) - min(counts$reference_date) + 1
+  if (missing(max_delay)) {
+    max_delay <- obs_delay
+    message(
+      "Using the maximum observed delay of ", max_delay, " days ",
+      "to complete the incidence data."
+    )
+  }
+  if (max_delay < obs_delay) {
+    message(
+      "Using the maximum observed delay of ", obs_delay,
+      " days as greater than the maximum specified to complete the incidence ",
+      "data.")
+       max_delay <- obs_delay
+  }
+  cum_counts <- enw_add_cumulative(counts, by = by, copy = FALSE)
+
+  complete_counts <- enw_complete_dates(
+    cum_counts, max_delay = max_delay, by = by,
+    completion_beyond_max_report = completion_beyond_max_report
+  )
+  complete_counts <- enw_add_incidence(complete_counts, by = by, copy = FALSE)
+  return(complete_counts[])
+}
+
+
+#' Convert Aggregate Counts (Incidence) to a Line List
+#'
+#' @description This function takes a `data.table` of aggregate counts or
+#' something coercible to a `data.table` (such as a `data.frame`) and converts
+#' it to a line list where each row represents a case.
+#'
+#' @param obs An object coercible to a `data.table` (such as a `data.frame`)
+#' which must have a `new_confirm` column.
+#'
+#' @param reference_date A character string of the variable name to use
+#' for the `reference_date` in the line list. If not supplied then the
+#' function will not rename the `reference_date` column.
+#'
+#' @param report_date A character string of the variable name to use
+#' for the `report_date` in the line list. If not supplied then the
+#' function will not rename the `report_date` column.
+#'
+#' @return A `data.table` with the following variables: `id`, `reference_date`,
+#' `report_date`, and any other variables in the `obs` object. Rows in `obs`
+#' will be duplicated based on the `new_confirm` column. `reference_date` and
+#' `report_date` may be renamed if `reference_date` and `report_date` are
+#' supplied.
+#'
+#' @export
+#' @family dataconverters
+#' @importFrom data.table setcolorder setnames
+#' @examples
+#' incidence <- enw_add_incidence(germany_covid19_hosp)
+#' incidence <- enw_filter_reference_dates(
+#'   incidence[location %in% "DE"], include_days = 10
+#' )
+#' enw_incidence_to_linelist(incidence, reference_date = "onset_date")
+enw_incidence_to_linelist <- function(obs, reference_date, report_date) {
+  obs <- coerce_dt(obs, dates = TRUE, required_cols = "new_confirm")
+  suppressWarnings(obs <- obs[, "confirm" := NULL])
+  cols <- setdiff(colnames(obs), "new_confirm")
+  obs <- obs[new_confirm > 0]
+  obs <- obs[, .(person = 1:new_confirm), by = cols]
+  obs[, person := NULL]
+  obs[, id := seq_len(.N)]
```

Comment:
seems a bit clunky; maybe just create/overwrite id, rather than create and destroy a column?